### PR TITLE
[Segment Replication] Add setting to enable segment replication as default replication type.

### LINKED ISSUE #6656: [Segment Replication] Provide option to turn on Segment Replication as default replication strategy.
Today, users need to enable segment replication by both setting a feature flag and index level setting while creating the index.
https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/segment-replication/configuration/

It would be ideal to provide an additional setting to pass via yml to make segrep the default replication strategy.

This setting should not replace the feature flag and should be an additional setting passed in opensearch.yml.  For now, it should only be enabled if the FF is also enabled.

### PR DESCRIPTION
**Pull request by Rishikesh1159**
### Description
This PR adds setting for users to enable segment replication as default replication strategy.

-> This setting is a static setting and can only be enable through configuration and not by REST API's.
-> When this setting is enabled it makes all new indices in cluster use segment replication as default replication type. But users can still create an index using document replication by passing replication type as DOCUMENT during index creation. As Index level settings always override cluster level settings.
-> Not only during index creation but also if users set `"index.replication.type"` in configuration file, as it is an index level setting it will override the new setting of default segment replication.
-> This setting is not applied for system indices. By default all system indices in opensearch will still use document replication even if this setting is enabled.
-> This setting is still behind a Segment Replication Feature flag. So, users have to still enable feature flag for this setting to work. Once feature flag is removed users can set this setting directly.

### Issues Resolved
Resolves #6656 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationDefaultIT.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+
+import static org.opensearch.cluster.metadata.IndexMetadata.DEFAULT_REPLICATION_TYPE_SETTING_SEGMENT;
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class SegmentReplicationDefaultIT extends OpenSearchIntegTestCase {
```
Comment (Poojita-Raj) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationDefaultIT.java:31]: SegmentReplicationDefaultIT doesn't tell us that this file is related to the setting. I would suggest renaming it to SegmentReplicationDefaultClusterSettingIT instead.

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationDefaultIT.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            .execute()
+            .actionGet();
+        if (isSystemIndex) {
+            // Verify that Segment Replication did not happen on the replica shard.
+            assertNull(segmentReplicationStatsResponse.getReplicationStats().get(indexName));
+        } else {
+            // Verify that Segment Replication happened on the replica shard.
+            assertFalse(segmentReplicationStatsResponse.getReplicationStats().get(indexName).get(0).getReplicaStats().isEmpty());
+        }
+    }
```
Comment (Poojita-Raj) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationDefaultIT.java:111]: We could add some more tests here to test the behavior described:

1. A test to ensure that when cluster setting for segrep is true, creating an index with docrep enabled overrides this cluster setting.
2. Similarly a test to ensure overriding takes place when index.replication.type is set to docrep when the segrep cluster setting is true. 
3. A last test (optional) to check that when the feature flag is not enabled, even setting the cluster setting doesn't result in segment replication indices.

File: server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java [hunk 3]
```diff
@@ -282,6 +282,16 @@ public Iterator<Setting<?>> settings() {
         Property.IndexScope
     );
 
+    /**
+     * Used to specify SEGMENT replication type as default for all indices. By default, this is false.
```
Comment (Poojita-Raj) [inline server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java:286]: Nit: 
```suggestion
     * Used to specify SEGMENT replication type as the default replication strategy for all indices in a cluster. By default, this is false.
```

File: server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java [hunk 4]
```diff
@@ -282,6 +282,16 @@ public Iterator<Setting<?>> settings() {
...
+     */
+    public static final Setting<Boolean> DEFAULT_REPLICATION_TYPE_SETTING_SEGMENT = Setting.boolSetting(
```
Comment (Poojita-Raj) [inline server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java:288]: Could be renamed to DEFAULT_SEGMENT_REPLICATION_TYPE_CLUSTER_SETTING or CLUSTER_DEFAULT_SEGMENT_REPLICATION_TYPE_SETTING to make it obvious this is a cluster level setting.

File: server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java [hunk 5]
```diff
@@ -282,6 +282,16 @@ public Iterator<Setting<?>> settings() {
...
+        "opensearch.index.default.replication.type.segment",
```
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java:289]: I would rename it to  something along the lines of `indices.replication.strategy` and set the default value to doc rep (preferably using an enum). The user can then set it to `seg_rep` as needed. 
I would also move it under `IndicesService` similar to - https://github.com/opensearch-project/OpenSearch/blob/f40fa82e3667825fd0ba8d4bb17b6959ff7cc4f1/server/src/main/java/org/opensearch/indices/IndicesService.java#L214-L219

**Comment by dreamer-89** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/12778/
> * **CommitID:** [7b3e333](https://github.com/opensearch-project/OpenSearch/commit/7b3e3339db706f29717eda609d5455ff183a3375)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Looks like a legit failure. 
```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.replication.SegmentReplicationDefaultIT.testReplicationWithDefaultSegmentReplicationSetting" -Dtests.seed=A834E113C4F1CF78 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=de-DE -Dtests.timezone=Australia/Queensland -Druntime.java=19

org.opensearch.indices.replication.SegmentReplicationDefaultIT > testReplicationWithDefaultSegmentReplicationSetting FAILED
    java.lang.AssertionError: expected null, but was:<[SegmentReplicationPerGroupStats{replicaStats=[SegmentReplicationShardStats{allocationId=s2L3gNsbQUei809hRc-pOg, checkpointsBehindCount=0, bytesBehindCount=0, currentReplicationTimeMillis=0, lastCompletedReplicationTimeMillis=45, currentReplicationState=org.opensearch.indices.replication.SegmentReplicationState@2e66ad62}], rejectedRequestCount=0}]>
        at __randomizedtesting.SeedInfo.seed([A834E113C4F1CF78:92D0663D70F53C93]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotNull(Assert.java:756)
        at org.junit.Assert.assertNull(Assert.java:738)
        at org.junit.Assert.assertNull(Assert.java:748)
        at org.opensearch.indices.replication.SegmentReplicationDefaultIT.testReplicationWithDefaultSegmentReplicationSetting(SegmentReplicationDefaultIT.java:106)
```

File: server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java [hunk 6]
```diff
@@ -317,22 +318,29 @@ public void validate(final Boolean value) {}
 
             @Override
             public void validate(final Boolean value, final Map<Setting<?>, Object> settings) {
+                final Object defaultClusterSettingReplicationType = settings.get(CLUSTER_DEFAULT_REPLICATION_TYPE_SETTING);
                 final Object replicationType = settings.get(INDEX_REPLICATION_TYPE_SETTING);
-                if (replicationType != ReplicationType.SEGMENT && value == true) {
+                if (replicationType != ReplicationType.SEGMENT
```
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/cluster/metadata/IndexMetadata.java:323]: nit: use `!ReplicationType.SEGMENT.equals(replicationType)` for safer usage

File: server/src/main/java/org/opensearch/indices/IndicesService.java [hunk 7]
```diff
@@ -230,6 +231,19 @@ public class IndicesService extends AbstractLifecycleComponent
         Setting.Property.NodeScope
     );
 
+    /**
+     * Used to specify SEGMENT replication type as the default replication strategy for all indices in a cluster. By default, this is false.
+     */
+    public static final String CLUSTER_SETTING_REPLICATION_TYPE = "cluster.indices.replication.strategy";
+
+    public static final Setting<ReplicationType> CLUSTER_DEFAULT_REPLICATION_TYPE_SETTING = new Setting<>(
```
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/indices/IndicesService.java:239]: `CLUSTER_REPLICATION_TYPE_SETTING` seems more appropriate here. Default is implied by the default value that will be provided. Also would need documentation around this to demonstrate the hierarchy of `CLUSTER_REPLICATION_TYPE_SETTING` and `INDEX_REPLICATION_TYPE_SETTING`

**Review by kotwanikunal — COMMENTED** [review]
Added a few more comments. 
Also add a `CHANGELOG` entry since you are introducing a new configuration.

File: server/src/main/java/org/opensearch/indices/IndicesService.java (hunk 7 above)
Comment (Rishikesh1159) [reply server/src/main/java/org/opensearch/indices/IndicesService.java:239]: Sure. I will update that. Yes, I will open an issue in documentation repo to document this changes once this PR is merged.

**Review by Poojita-Raj — APPROVED** [review]

**Review by kotwanikunal — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/IndexSettings.java [hunk 8]
```diff
@@ -750,7 +751,13 @@ public IndexSettings(final IndexMetadata indexMetadata, final Settings nodeSetti
         nodeName = Node.NODE_NAME_SETTING.get(settings);
         this.indexMetadata = indexMetadata;
         numberOfShards = settings.getAsInt(IndexMetadata.SETTING_NUMBER_OF_SHARDS, null);
-        replicationType = ReplicationType.parseString(settings.get(IndexMetadata.SETTING_REPLICATION_TYPE));
+        if (FeatureFlags.isEnabled(FeatureFlags.REPLICATION_TYPE)
+            && indexMetadata.isSystem() == false
+            && settings.get(IndexMetadata.SETTING_REPLICATION_TYPE) == null) {
+            replicationType = ReplicationType.parseString(settings.get(IndicesService.CLUSTER_SETTING_REPLICATION_TYPE));
```
[the remark this answers is not in the document]
Comment (Rishikesh1159) [reply server/src/main/java/org/opensearch/index/IndexSettings.java:757]: Sure, I will update this

**Comment by Rishikesh1159** [comment]
Currently, when user set cluster setting `"cluster.indices.replication.strategy"` in opensearch.yml, all the newly created indices will use segment replication as default replication strategy. But when user tries to fetch index settings `GET /<index>/_settings` the response does not show replication type as SEGMENT. So this is a bug and it might be confusing for users. I am currently looking for any other workarounds to fix this issue, I will update here as soon as I find something. Any other thoughts are much appreciated please.

**Pull request merged**